### ISSUE TITLE
Feature Request: builtin ACME client support

### ISSUE THREAD

**Issue by akvadrako** [haproxy/haproxy#1864]
### Your Feature Request

HAProxy does more of what I want than competing reverse proxies, but the thing that makes me sometimes reach for Caddy or traefik is the built-in ACME (Let's Encrypt) client. It's very convenient to have SSL certificates automatically managed, especially in two use cases:

1. When you have lots of domains behind one proxy, and often add/remove them.
2. For quick experiments.

In Caddy, you only need to specify your domain name and add your email address. Versus:

1. running and configuring a certbot container
2. writing a script which prepares those certificates and reloads the proxy
3. creating a redirect rule for each domain
4. creating a backend for the certbot

I'm aware of the `haproxy-lua-acme` module, but it's not as good as the above steps. It's hasn't been maintained for almost 2 years and it requires some packaging work. So it would be nice if this was handled in a more clean and supported way by the main distribution.

### What are you trying to do?

Spend less time managing SSL certificates.

### Output of `haproxy -vv`

```plain
HAProxy version 2.6.5-987a4e2 2022/09/03 - https://haproxy.org/
```

Edit:

Merged as experimental in 3.2.

**Comment by lukastribus** [comment]
While "nice to have", this presents some difficulties that by design are not easy to solve in haproxy:

- we cannot and do not want to write to disk
- we will never be able to write outside chroot and or with privileges that we already dropped

So I'm not sure how realistic it is to build an ACME client into haproxy with full functionality. Likely an external helper will be needed.

**Comment by akvadrako** [comment]
If the certificates are cached in memory between reloads, disk persistence is not strictly necessary. Let's Encrypt allows 50 requests per TLD per week, though if you exceed that it would cause trouble.

An external helper that provides the disk persistence sounds like a good idea, if it's simple enough. Like just a docker one-liner and that can be enabled with a single configuration directive. Perhaps something akin to the Global Profiling Engine.

**Comment by git001** [comment]
As it sounds like a good Idea it's not a easy task :smile: .

As I'm a fan of the [KISS Principle](https://en.wikipedia.org/wiki/KISS_principle) my current workflow is this and I'm quite sure similar steps are out there to get the certificates and key updated.

```
certbot renew
cat /etc/letsencrypt/live/<DOMAIN>/fullchain.pem \
  /etc/letsencrypt/live/<DOMAIN>/privkey.pem \
  > /usr/local/etc/haproxy/stream-cert.pem
systemctl restart haproxy.service
```

The haproxy.service looks like this.

```systemctl cat haproxy.service
# /etc/systemd/system/haproxy.service
[Unit]
Description=haproxy Podman in Systemd

[Service]
Restart=on-failure
ExecStop=/usr/bin/sh -c "/usr/bin/podman rm -f `cat /%t/%n-cid`"
ExecStartPre=/usr/bin/rm -f /%t/%n-pid /%t/%n-cid
ExecStart=/usr/bin/podman run --name haproxy --rm \
  --network host \
  --publish 80:80 \
  --publish 443:443 \
  --publish 5555:5555 \
  --publish 8404:8404 \
  --publish 8405:8405 \
  --publish 2020:2020 \
  --publish 1936:1936 \
  --conmon-pidfile  /%t/%n-pid \
  --cidfile /%t/%n-cid -d \
  -v /usr/local/etc/haproxy:/usr/local/etc/haproxy:Z \
  haproxytech/haproxy-ubuntu
KillMode=none
Type=forking
PIDFile=/%t/%n-pid

[Install]
WantedBy=multi-user.target
```

From my point of view would it be nice to use the runtime api to update the certificates after the certbot have renewed the certificates.
I prefer for that task the cerbots as there are several plugins for DNS Providers which makes the maintenance quite easy.

**Comment by wlallemand** [comment]
There are currently some development about this, but that requires some major change of architecture. We already spoke about a separate process (like for the program section) that could access the files for writing like suggested, that's one of the lead we have for a acme implementation.

**Comment by FedericoCeratto** [comment]
https://github.com/dehydrated-io/dehydrated provides a client (orders of magnitude simpler than certbot).
Both haproxy and dehydrated can by ran with limited privileges using SystemD without architecture changes. See e.g. https://github.com/dehydrated-io/dehydrated/pull/893/files
One way of integrating them is https://github.com/LucaFilipozzi/til/blob/main/haproxy/responding_to_acme_challenges_with_haproxy.md
This could be the basis to create a simple ACME tool that runs besides haproxy and with very limited privileges.

**Comment by wlallemand** [comment]
@FedericoCeratto there are already plenty of solutions which use a separated client indeed, certbot, acme.sh etc  which works well once correctly configured. But this ticket is about a builtin ACME client which would be integrated into the haproxy configuration.

**Comment by wlallemand** [comment]
quick note about this, a builtin ACME client won't be released for now, however I made some changes to the deployment script of acme.sh to be able to use the stats socket to update or add new certificates.

the PR is still pending there: https://github.com/acmesh-official/acme.sh/pull/4581

But we integrated the deploy script in the haproxy repository, so you can follow this to deploy:
https://github.com/haproxy/wiki/wiki/Letsencrypt-integration-with-HAProxy-and-acme.sh

acme.sh is also able to run in stateless mode, which allows HAProxy to serve the ACME challenge easily with just a configuration line:
https://github.com/acmesh-official/acme.sh/wiki/Stateless-Mode/#haproxy

with these 2 features you are able to run acme.sh in non-root and without the need to serve the challenge from acme.sh

**Comment by lukastribus** [comment]
The TLS ALPN method allows to serve acme.sh in non-root mode also, which I personally like very much (no need to insert static account thumbprints in haproxy configuration):

https://github.com/acmesh-official/acme.sh/wiki/TLS-ALPN-without-downtime

**Comment by markkrj** [comment]
@lukastribus  but using the ALPN method, every connection would be proxied to localhost... Wouldn't it cause latency implications? Even if minimal. At least, one more TCP handshake.

**Comment by lukastribus** [comment]
Yes, it is a second proxying layer, so the impact is necessarily non-zero, but really I doubt that this is a practical problem.

Here's a post of Willy Tarreau from 15 years ago:

https://linux-kernel.vger.kernel.narkive.com/L5ljiEVb/unix-sockets-via-tcp-on-localhost-is-tcp-slower#post15

To avoid the TCP handshake latency penalty over loopback despite using a 2 layer proxying approach you can use unix sockets or even better, abstract namespace sockets (so you avoid dealing with path and permission issues).


If we would build a native ACME client within haproxy, I would suggest that we would do RFC8737 aka "TLS-ALPN-01" as primary method, because it doesn't require an exposed port 80.

I guess it's more complicated then the HTTP challenge though, as it requires a little more TLS magic than just a different ALPN negotiation. Considering that at some point we are probably going to support wolfssl or some other openssl alternative, this adds additional work.

Caddy has this:
https://caddy.community/t/caddy-supports-the-acme-tls-alpn-challenge/4860


Lukas

**Comment by wtarreau** [comment]
I didn't know about the ALPN approach, it's interesting. The only thing I don't like is relaying to a local port, leaving the acme server exposed (or the port opened to anything that wants to take it). Maybe an intermediate solution would be to put the response challenge in an environment variable so that it doesn't appear in the config itself.

**Comment by wlallemand** [comment]
> Maybe an intermediate solution would be to put the response challenge in an environment variable so that it doesn't appear in the config itself.

The configuration example I wrote for the stateless mode is already doing it this way, it uses setenv in the global section but could be set from the unit file or elsewhere. It's not much of a security issue, you need the client to perform the request with the account key, and only the client gets the signed certificate, it's just another config entry. 

Regarding the TLS-ALPN-01 challenge, it's indeed implemented by acme.sh, but I honestly prefer the stateless method which does not require to route to acme.sh as a server. But indeed this is useful if you don' t want the port 80 exposed. What we could try though, is to implement the TLS-ALPN-01 method in stateless mode and add a way to send the response-challenge from haproxy with the alpn.

**Comment by lukastribus** [comment]
The ALPN routing to a local configuration is a intermediate solution.

The acme server however only runs (and therefore listens) during a verification; the local port could be moved to below 1024 to restrict this to root user; so non privileged users cannot listen to this port by mistake.

If on the other hand we are talking about a native implementation in haproxy, the TLS-ALPN-01 is nice because it doesn't require port 80 indeed.

**Comment by markkrj** [comment]
@wlallemand hmm, a native solution would be awesome, something like "option challenge-tls-alpn-01" :)
But I guess generating the cert would be a problem, because HAProxy does not write to disk... But as certs are short lived, just for checking the challenge, they could reside in memory, no?

**Comment by git001** [comment]
Well that could be a problem for HA-Setups. The Certificate must be transferred to all HAProxy Instances which makes it quite complicated if the Certificate is only in memory. One option could be to send the Certificate after it was created via peers to the other Instances, but this works only if the Instances are connected via peers, another option could be to get the Certificate via HAProxy socket commands or create a LUA script which distributes the Certificate to the other Instances, as you can see that's a not trivial task.

**Comment by wlallemand** [comment]
Let's resume:

- TLS-ALPN-01 method is best suited when you don't want to have the port 80 exposed. Currently only doable with acme.sh + haproxy by proxying to the acme.sh server. No "stateless" mode with acme.sh is possible for now with HAProxy.
If we could at least implement this for the stateless acme.sh mode it would be a big step and will probably be reused for a builtin ACME client in haproxy.

- HTTP-01 method is currently doable in stateless mode with acme.sh + haproxy without the need to proxify to acme.sh, but it requires the port 80. It is useful in environment where the ACME client and the server is not on the same machine.

- If we implement a builtin ACME client, there will be access to the disk indeed. Don't focus on the current architecture where the *worker* does not have access to the disk, the master is able to launch multiple processes. Also communication between processes is possible using the internal socketpair. In fact that would be kind of the same thing that acme.sh is implementing with the deploy script: acme.sh is writing to the disk, and then update the certificate over the CLI.

- Regarding the update of certificates on multiple instances it's another subject which is unrelated to ACME, it's a problem already present with certificates that people usually fix with a certificate vault and deployment scripts. We don't have any builtin solutions though.

**Comment by lukastribus** [comment]
Slightly off-topic, but jet another reason one may want to use TLS-ALPN-01 as opposed to HTTP-01: command injection attacks due to acme client bugs are actually a thing: https://github.com/acmesh-official/acme.sh/issues/4659

I think this would be way harder with TLS-ALPN-01.

edit: wrong. The vulnerability is not in the challenge phase, but in the acme interaction between client and CA, so this can happen with all challenge types, not only HTTP-01.

**Comment by ndilieto** [comment]
https://github.com/ndilieto/uacme?tab=readme-ov-file#tls-alpn-01-challenge-support

> [ualpn](https://ndilieto.github.io/uacme/ualpn.html) is a lightweight proxying tls-alpn-01 challenge responder, designed to handle incoming HTTPS connections on port 443. Most of the time it just transparently proxies connections to the real web server (which can be on either another machine, or a different TCP port on the same machine). When a tls-alpn-01 challenge handshake comes in ualpn handles it on the fly instead of proxying it to the webserver. This means that unlike other available tls-alpn-01 responders, ualpn does not require your webserver to stop during the challenge (zero downtime).

**Comment by wlallemand** [comment]
Update on this, we started to write a builtin ACME client which supports HTTP-01 for now. This is a work in progress and will be available as an experimental feature for version 3.2. We still have restrictions on the filesystem from within haproxy so saving the files will require external tooling.

The documentation was merged in https://github.com/haproxy/haproxy/commit/f9390a689f234f360fbabcd81982e6f61ec5bd48 but this will still evolve before the release.

**Comment by nakoo** [comment]
> Update on this, we started to write a builtin ACME client which supports HTTP-01 for now. This is a work in progress and will be available as an experimental feature for version 3.2. We still have restrictions on the filesystem from within haproxy so saving the files will require external tooling.
> 
> The documentation was merged in [f9390a6](https://github.com/haproxy/haproxy/commit/f9390a689f234f360fbabcd81982e6f61ec5bd48) but this will still evolve before the release.

The HAProxy runtime doesn't support this feature in 3.2-dev11 yet, does it?

**Comment by wlallemand** [comment]
@nakoo this was merged just before 3.2-dev11 so this version will support it,  but more patches and fixes are coming soon.

**Comment by nakoo** [comment]
> [@nakoo](https://github.com/nakoo) this was merged just before 3.2-dev11 so this version will support it, but more patches and fixes are coming soon.

Oh, it seems the `acme renew` command is available, but the `help` command has not been updated yet.

**Comment by wlallemand** [comment]
@nakoo right, I hid it  from help during development, I'll add it to the help again.

**Comment by wlallemand** [comment]
3.2 will be released around the end of the month, don't hesitate to check the latest dev version, any feedback would be appreciated! dev14 now has a scheduler so you don't need to start manually renewals.

**Comment by nakoo** [comment]
I tested version 3.2-dev-14, but I'm not sure I set it up correctly.
When I run `acme ps` or `acme renew web.crt` in haproxy-runtime, nothing appears.

> haproxy.cfg
```
crt-store web
    crt-base /run/secrets
    key-base /run/secrets

    load crt "web.crt"     key "web.key"     alias "web"     acme LE domains "web.example.com"

acme LE
    directory https://acme-v02.api.letsencrypt.org/directory
    account-key /run/secrets/account.key
    curves P-384

frontend web
    bind quic4@:443 ssl strict-sni crt "@web/web" alpn h3
    bind *:443      ssl strict-sni crt "@web/web" tls-ticket-keys /run/secrets/tls-ticket.key alpn h2,http/1.1
```

> acme ps
```bash
# certificate	section	state
```

> acme renew web.crt
```bash
Can't find the certificate 'web.crt'.
```

**Comment by wlallemand** [comment]
@nakoo In your case your certificate name is "@web/web" as shown in "show ssl cert". The CLI command does not take the filename when you specify an alias, same as the 'crt' option.

But with dev14 you don't need to use the 'renew' command anymore if you don't want to force the renewal manually, this will be started when needed.

Take a look at your traffic log if you want to see the ACME HTTP requests, and your admin logs will show details. You can then check your certificate with "show ssl cert" to see details.

Also I removed the 'acme ps' command and replaced it by 'acme status', which is less confusing and will show every certificates that are configured for acme, not only the current task. (in the current master branch)

**Comment by nakoo** [comment]
Thank you. I could test with `acme renew @web/web`. 

When I ran the command, it returned an "out of memory" message after a few seconds.
However, the renewal process continued but in the end resulted in an error.

> acme renew @web/web
```bash
Out of memory.
```

> haproxy.log
```bash
acme: @web/web: Starting update of the certificate.
-:- [06/May/2025:15:33:00.266] <ACME> -/- 7/0/323/317/645 200 1306 - - ---- 1/0/0/0/0 0/0 {172.65.XX.XXX} "GET https://acme-staging-v02.api.letsencrypt.org/directory HTTP/1.1" 0/0000000000000000/-/-/0 -/-/-
-:- [06/May/2025:15:33:00.912] <ACME> -/- 2/0/0/159/159 200 309 - - ---- 1/0/0/0/0 0/0 {172.65.XX.XXX} "HEAD https://acme-staging-v02.api.letsencrypt.org/acme/new-nonce HTTP/1.1" 0/0000000000000000/-/-/0 -/-/-
-:- [06/May/2025:15:33:01.072] <ACME> -/- 4/0/0/171/173 200 769 - - ---- 1/0/0/0/0 0/0 {172.65.XX.XXX} "POST https://acme-staging-v02.api.letsencrypt.org/acme/new-acct HTTP/1.1" 0/0000000000000000/-/-/0 -/-/-
-:- [06/May/2025:15:33:01.245] <ACME> -/- 4/0/0/198/200 201 822 - - ---- 2/0/0/0/0 0/0 {172.65.XX.XXX} "POST https://acme-staging-v02.api.letsencrypt.org/acme/new-order HTTP/1.1" 0/0000000000000000/-/-/0 -/-/-
-:- [06/May/2025:15:33:01.445] <ACME> -/- 4/0/0/170/172 200 1217 - - ---- 2/0/0/0/0 0/0 {172.65.XX.XXX} "POST https://acme-staging-v02.api.letsencrypt.org/acme/authz/19846XXXX/1721540XXXX HTTP/1.1" 0/0000000000000000/-/-/0 -/-/-
-:- [06/May/2025:15:33:01.618] <ACME> -/- 4/0/0/170/172 200 765 - - ---- 2/0/0/0/0 0/0 {172.65.XX.XXX} "POST https://acme-staging-v02.api.letsencrypt.org/acme/chall/19846XXXX/1721540XXXX/gYXsfg HTTP/1.1" 0/0000000000000000/-/-/0 -/-/-
acme: @web/web: challenge error: "The key authorization file from the server did not match this challenge. Expected "q0H2OIUitUr33lBiruhvZaGiVFEeCooaa-2CZAnyMAk.kdb_0SLUbSdiLrZcHJCnBzdKNnOsHAcHjAKYViDf3As" (got "")" (urn:ietf:params:acme:error:unauthorized) (HTTP status code 200), aborting.
-:- [06/May/2025:15:33:06.792] <ACME> -/- 2/0/0/163/163 200 1311 - - ---- 2/0/0/0/0 0/0 {172.65.XX.XXX} "GET https://acme-staging-v02.api.letsencrypt.org/acme/chall/19846XXXX/1721540XXXX/gYXsfg HTTP/1.1" 0/0000000000000000/-/-/0 -/-/-
```

I think the challenge URL returns an empty string because I set my `default_backend` like this, but I'm not sure if acme option requires a special backend for this.

```haproxy
backend empty
    http-request return status 200
```

I think I also need to mention that I am currently using the `haproxy-docker-alpine-quic` docker image for this.

**Comment by wlallemand** [comment]
> When I ran the command, it returned an "out of memory" message after a few seconds.

Indeed, this was already fixed in the master, but that's only a wrong messsage.


> I think the challenge URL returns an empty string because I set my `default_backend` like this, but I'm not sure if acme option requires a special backend for this.

Please take a look at the example there: https://docs.haproxy.org/dev/configuration.html#3.13 . You must specify the "http-request return" rule which use a map from your acme section so it could reply the challenge correctly.

**Comment by nakoo** [comment]
Yes, after adding the line, it seems haproxy returns the challenge code. 
However, the part `kdb_0SLUbSdiLrZcHJCnBzdKNnOsHAcHjAKYViDf3As` is missing, and an error is returned.


```json
{
  "type": "http-01",
  "url": "https://acme-staging-v02.api.letsencrypt.org/acme/chall/19846XXXX/1721635XXXX/ClwZeQ",
  "status": "invalid",
  "validated": "2025-05-06T16:36:11Z",
  "error": {
    "type": "urn:ietf:params:acme:error:unauthorized",
    "detail": "The key authorization file from the server did not match this challenge. Expected \"JmoLet2wY-8-vpjfeqtUAt8lsqju7-WAnz0DQUguxdo.kdb_0SLUbSdiLrZcHJCnBzdKNnOsHAcHjAKYViDf3As\" (got \"JmoLet2wY-8-vpjfeqtUAt8lsqju7-WAnz0DQUguxdo.\")",
    "status": 403
  },
  "token": "JmoLet2wY-8-vpjfeqtUAt8lsqju7-WAnz0DQUguxdo",
  "validationRecord": [
    {
      "url": "http://web.example.com/.well-known/acme-challenge/JmoLet2wY-8-vpjfeqtUAt8lsqju7-WAnz0DQUguxdo",
      "hostname": "web.example.com",
      "port": "80",
      "addressesResolved": [
        "193.123.XXX.XXX"
      ],
      "addressUsed": "193.123.XXX.XXX"
    }
  ]
}
```

**Comment by nakoo** [comment]
Oh, I overlooked `map virt@acme`. After adding that line, it started working.

**Comment by NickMRamirez** [comment]
To keep the separation of concerns separate, I think that `ssl-f-use` should not duplicate arguments that the `load` directive owns under `crt-store`. My theory is that:

* `crt-store` manages certificate and key locations.
* `ssl-f-use` manages references to `crt-store` files and crt-list type properties.

Where this is maintaining the separation:

```
acme letsencrypt-staging
  directory https://acme-staging-v02.api.letsencrypt.org/directory
  account-key /etc/haproxy/account.key
  contact admin@example.com
  challenge HTTP-01
  keytype RSA
  bits 2048
  map virt@acme

crt-store my_files
  crt-base /etc/haproxy/
  key-base /etc/haproxy/
  load crt "example.com.pem" acme letsencrypt-staging domains "example.com" alias "example"

frontend mysite
  bind :80
  bind :443 ssl
  http-request return status 200 content-type text/plain lf-string "%[path,field(-1,/)].%[path,field(-1,/),map(virt@acme)]\n" if { path_beg '/.well-known/acme-challenge/' }
  http-request redirect scheme https unless { ssl_fc }
  ssl-f-use crt "@my_files/example" 
  use_backend webservers
```

But this doesn't maintain the separation because `ssl-f-use` uses  a direct location of a file instead of a crt-store reference, and `acme` and `domain` arguments, which better belong to `load`:

```
frontend mysite
  bind :80
  bind :443 ssl
  http-request return status 200 content-type text/plain lf-string "%[path,field(-1,/)].%[path,field(-1,/),map(virt@acme)]\n" if { path_beg '/.well-known/acme-challenge/' }
  http-request redirect scheme https unless { ssl_fc }
  ssl-f-use crt "example.com.pem" acme letsencrypt-staging domains "example.com"
  use_backend webservers
```

What's I'm suggesting is that `ssl-f-use` must refer to a `crt-store` loaded certificate to maintain "storage  / location" and "usage" as separate. And here "location" is Let's Encrypt, so I think all ACME details belong under crt-store.

**Comment by nakoo** [comment]
I also want an option to disable the ACME scheduler because I might need to run `dump ssl cert` when HAProxy renews the certificates, and want to avoid  unexpected surprises.

**Comment by wlallemand** [comment]
> To keep the separation of concerns separate, I think that `ssl-f-use` should not duplicate arguments that the `load` directive owns under `crt-store`. My theory is that:
> 
>     * `crt-store` manages certificate and key locations.

The crt-store is about configuring the storage, and everything related to the storage, so no SSL options there, only things about the PEM, OCSP, ACME etc. It only declares the storage but does not initialize it. Before the creation of the "crt-store", this storage was already existing, any "crt" or certificate from a crt-list would fill this storage if the certificate was not found into it before initializing it.

>     * `ssl-f-use` manages references to `crt-store` files and crt-list type properties.

ssl-f-use is a crt-list line, so like a crt-list line it would take storage options and sslbindconf options. So you could load a certificate directly from the ssl-f-use option.
The whole point about ssl-f-use, is to fix the configuration problem that we have with "crt" on a bind line, which does not permit to configure a "key" or storage options for multiple certificates. Implementing the crt-store was a first step to allow this, but it would require a separated section, which is not useful for light configuration.

> Where this is maintaining the separation:
> 
> ```
> acme letsencrypt-staging
>   directory https://acme-staging-v02.api.letsencrypt.org/directory
>   account-key /etc/haproxy/account.key
>   contact admin@example.com
>   challenge HTTP-01
>   keytype RSA
>   bits 2048
>   map virt@acme
> 
> crt-store my_files
>   crt-base /etc/haproxy/
>   key-base /etc/haproxy/
>   load crt "example.com.pem" acme letsencrypt-staging domains "example.com" alias "example"
> 
> frontend mysite
>   bind :80
>   bind :443 ssl
>   http-request return status 200 content-type text/plain lf-string "%[path,field(-1,/)].%[path,field(-1,/),map(virt@acme)]\n" if { path_beg '/.well-known/acme-challenge/' }
>   http-request redirect scheme https unless { ssl_fc }
>   ssl-f-use crt "@my_files/example" 
>   use_backend webservers
> ```
> 
> But this doesn't maintain the separation because `ssl-f-use` uses a direct location of a file instead of a crt-store reference, and `acme` and `domain` arguments, which better belong to `load`:
> 

I'm not sure I'm following, this configuration is valid, ssl-f-use, like any crt keyword or crt-list, takes a certificate identifier in parameter, historically this identifier was the filename, but with the introduction of the aliases, it could be an alias starting by "@". So it first tries  to look for a crt-store entry which was previously filed, and if not, it would try to load a filename instead.

> ```
> frontend mysite
>   bind :80
>   bind :443 ssl
>   http-request return status 200 content-type text/plain lf-string "%[path,field(-1,/)].%[path,field(-1,/),map(virt@acme)]\n" if { path_beg '/.well-known/acme-challenge/' }
>   http-request redirect scheme https unless { ssl_fc }
>   ssl-f-use crt "example.com.pem" acme letsencrypt-staging domains "example.com"
>   use_backend webservers
> ```
> 
> What's I'm suggesting is that `ssl-f-use` must refer to a `crt-store` loaded certificate to maintain "storage / location" and "usage" as separate. And here "location" is Let's Encrypt, so I think all ACME details belong under crt-store.

Both are supported, using ssl-f-use with the storage options is an easy way to configure a simple setup, where you don't want to use your certificate at multiple places, it's like adding a new "crt" on the bind line, but more flexible. However the crt-store was designed to configure the store for certificates once for all at the same place when you want to use the certificate at multiple places so you only refer to it.

But I don't want to force people to use a crt-store if that's what you are suggestingg, that does not make sense to me, it's more complicated to write a configuration this way if you only have a simple setup.

**Comment by wlallemand** [comment]
> I also want an option to disable the ACME scheduler because I might need to run `dump ssl cert` when HAProxy renews the certificates, and want to avoid unexpected surprises.

I agree, I planned to add it for this release.

**Comment by wlallemand** [comment]
@NickMRamirez to be clearer, these 2 configurations are working:

```
crt-store
    load crt "example.crt" key "example.key" acme LE domains "example.com" alias "site1"

frontend in1
    bind *:4443 ssl
    ssl-f-use crt "@/site1"

acme LE
    directory https://acme-staging-v02.api.letsencrypt.org/directory
    contact admin@example.com
```

OR

```
frontend in2
    bind *:4443 ssl
    ssl-f-use  crt "example.crt" key "example.key" acme LE domains "example.com"
acme LE
    directory https://acme-staging-v02.api.letsencrypt.org/directory
    contact admin@example.com
```

And for a small setup you probably want the 2nd one.

**Comment by NickMRamirez** [comment]
Yes, both are working.

**Comment by wlallemand** [comment]
@nakoo I just pushed `acme.scheduler off` global option in master.

**Comment by gtags** [comment]
What is the proper way to create the first certificate using a new buildin ACME client?

If I already have a certificate, I can renew it and load it via the HAProxy Runtime API. But I couldn’t find a CLI command like `acme create` to get the initial certificate.

When I try to reference a non-existent cert file using `load crt ... acme LE ...` or `ssl-f-use crt ... acme LE ...`, the config parser fails with:

```
config : unable to stat SSL certificate from file 'example.com.pem' : No such file or directory.
```
Any tips or clarification would be much appreciated—thanks!

**Comment by wlallemand** [comment]
@gtags that's one of the limitation for now, access to the disk is not possible once the configuration is loaded, and we don't generate these at startup. In the future it will be possible to start without a filename or to specify a file that does not exist yet.

But in the current state of this feature, you need an existing keypair on the filesystem when starting, you could just generate a self-signed certificate like this:

```
openssl req -x509 -nodes -newkey ec -pkeyopt ec_paramgen_curve:secp384r1 -keyout example.com.pem -out example.com.pem
```

**Comment by zhangyoufu** [comment]
Does built-in ACME client support choosing [certificate profile](https://letsencrypt.org/2025/01/09/acme-profiles/)? It is mandatory for [shortlived certificates and IP certificates](https://letsencrypt.org/2025/01/16/6-day-and-ip-certs/) which ought to be GA in late 2025.

Let's Encrypt provides [three kinds of profiles](https://letsencrypt.org/docs/profiles/). `shortlived` profile is currently under allowlist, but [should be available in the staging environment "soon"](https://community.letsencrypt.org/t/does-lets-encrypt-support-ip-certificates/238155/3).

**Comment by wlallemand** [comment]
@zhangyoufu not yet, we are focusing on finalizing features from the current RFC 8555 before implementing things that are not released yet as an RFC.

**Comment by zhangyoufu** [comment]
Could you please use lowercase `http-01` and `dns-01` as the identifier of the challenge validation method, instead of uppercase `HTTP-01` and `DNS-01`?

**Comment by wlallemand** [comment]
@zhangyoufu why would you need that? I know that's it is experimental but I don't want to introduce configuration changes that are not necessary.

**Comment by zhangyoufu** [comment]
> [@zhangyoufu](https://github.com/zhangyoufu) why would you need that? I know that's it is experimental but I don't want to introduce configuration changes that are not necessary.

Both RFC8555 and [IANA registry](https://www.iana.org/assignments/acme/acme.xhtml#acme-validation-methods) uses lowercase, so I think it would be better to stay lowercase.

Could we change the document and default value to lowercase, and make this configuration accept both lowercase/uppercase value?

**Comment by zhangyoufu** [comment]
I'm interested in `tls-alpn-01` method. To implement `tls-alpn-01` method, we have limited choices on how to integrate it with SSL-enabled frontend (instead of the traditional `tcp-request inspect-delay` and `req.ssl_alpn` way):

- Should we adjust `crt-list` format?
  `crt-list` has the following format for each line: `<crtfile> [\[<sslbindconf> ...\]] [[!]<snifilter> ...]`.
  Where `<crtfile>` and `<sslbindconf> ...` are the "action" part and `<snifilter> ...` are the "match" part.
  How about `<alpnfilter> ...` at the tail of each line?
  How to design the syntax so that parser could differentiate between `<alpnfilter> ...` and `<snifilter> ...`?
  SNI is single-value by nature, but ALPN is a list. How to design the ALPN match?
- Should we give up matching SNI and ALPN together?
  Should we match ALPN before `crt-list` and `<snifilter> ...`?
  Something like `crt-list <file> [{if | unless} <condition>]` can make use of `ssl_fc_alpn` sample and offers great flexibility.
  (It's questionable whether this flexibility is really needed.)
  The built-in acme client could maintain separate crt-list consists of self-signed ALPN certificates.
- Should we give up general ALPN filter capability?
  Should we implement `tls-alpn-01` (ALPN `acme-tls/1`) as a special case without introducing any configuration change?

What do you think?

**Comment by wlallemand** [comment]
> > [@zhangyoufu](https://github.com/zhangyoufu) why would you need that? I know that's it is experimental but I don't want to introduce configuration changes that are not necessary.
> 
> Both RFC8555 and [IANA registry](https://www.iana.org/assignments/acme/acme.xhtml#acme-validation-methods) uses lowercase, so I think it would be better to stay lowercase.
> 
> Could we change the document and default value to lowercase, and make this configuration accept both lowercase/uppercase value?

Well that's a good point, if we need to introduce more challenges it will be better to use the IANA naming, I'll make it compatible with both.

**Comment by wlallemand** [comment]
> I'm interested in `tls-alpn-01` method. To implement `tls-alpn-01` method, we have limited choices on how to integrate it with SSL-enabled frontend (instead of the traditional `tcp-request inspect-delay` and `req.ssl_alpn` way):
> 

I didn't work on tls-alpn-01 yet, but I was thinking of doing a tcp-request rule to reply to the challenge.

>     * Should we adjust `crt-list` format?
>       `crt-list` has the following format for each line: `<crtfile> [\[<sslbindconf> ...\]] [[!]<snifilter> ...]`.
>       Where `<crtfile>` and `<sslbindconf> ...` are the "action" part and `<snifilter> ...` are the "match" part.
>       How about `<alpnfilter> ...` at the tail of each line?
>       How to design the syntax so that parser could differentiate between `<alpnfilter> ...` and `<snifilter> ...`?
>       SNI is single-value by nature, but ALPN is a list. How to design the ALPN match?
> 
>     * Should we give up matching SNI and ALPN together?
>       Should we match ALPN before `crt-list` and `<snifilter> ...`?
>       Something like `crt-list <file> [{if | unless} <condition>]` can make use of `ssl_fc_alpn` sample and offers great flexibility.
>       (It's questionable whether this flexibility is really needed.)
>       The built-in acme client could maintain separate crt-list consists of self-signed ALPN certificates.
> 
>     * Should we give up general ALPN filter capability?
>       Should we implement `tls-alpn-01` (ALPN `acme-tls/1`) as a special case without introducing any configuration change?
> 
> 
> What do you think?

That does not seem like a good idea to implement more filtering in the crt-list, we already suffers from some inconsistencies with TLS options that could be set on both the bind line and a crt-list line, and people tend to confuse options with filtering. The filtering is only done for SNI and ECDSA/RSA selection. I don't think we will introduce more complexity there.


In my opinion we just need a rule to reply to the challenge, like it's done for http-01. We could imagine to introduce a new type of rules for clienthello matching on SSL-enabled frontend.

**Comment by zhangyoufu** [comment]
@wlallemand The cli command usage message for `acme challenge_ready` seems to be copied from `acme status`. Both `show status of certificates configured with ACME`. This needs to be fixed.

https://github.com/haproxy/haproxy/blob/8fb5ae5cc66259a616de736d94749b63acda3d4d/src/acme.c#L2744-L2745

**Comment by wlallemand** [comment]
@zhangyoufu thanks, this was fixed in commit fbffd2e25f2ec2333bf70f018d5617d62360c2ab

**Comment by wlallemand** [comment]
I created a new wiki page regarding the native implementation: https://github.com/haproxy/wiki/wiki/ACME:--native-haproxy

**Comment by hpn0t0ad** [comment]
I really like this new feature and it works nicely with Let’s Encrypt. Any chance EAB support could be added? Nearly all other ACME-compatible CAs I know of seem to require it.

**Comment by wlallemand** [comment]
@hpn0t0ad that's in my TODO list!

**Comment by wtarreau** [comment]
Stupid question: shouldn't we close this one now that the feature is implemented, and create new issues only for new feature requests ? Otherwise I fear that this issue will become a discussion forum about everything related to ACME, both bugs and feature requests.

**Comment by git001** [comment]
+1 for close

**Comment by wlallemand** [comment]
Indeed, closing.

**Comment by Ra2-IFV** [comment]
I'm getting error with example setup in docs:

```text
acme: @example/alias: Starting update of the certificate.
acme: @example/alias: Starting update of the certificate.
[ALERT]    (179342) : httpclient: SSL was disabled (wrong verify/ca-file)!
[ALERT]    (179342) : httpclient: Failed to initialize appctx httpclient_start:360.
acme: @example/alias: , retrying in 3s (1/5 retries)...
acme: @example/alias: , retrying in 3s (1/5 retries)...
```

Config:

```
load crt 'null.cert.name' domains 'example.com' alias 'alias' acme LE1
```

Is this feature ready for testing?

**Comment by wlallemand** [comment]
@Ra2-IFV  Your httpclient does not seem to be working correctly, you might lack the ca-certificates on your setup. Please open a ticket with details on your setup! Thanks!

**Comment by Ra2-IFV** [comment]
Thank you, just want some quick sanity check, I'll open a issue.
